### DISCUSSION TITLE
direct access to memory buffer

### DISCUSSION THREAD

**Discussion by stryxter** [olikraus/u8g2#2138] [Q&A] [answered]
is there a way i can access the memory buffer directly (read/write) and access the specific pixels in the buffer. For example i want to read if there is a pixel set at the coordinates (10,55) of my ST7920 128x64 graphical lcd

(more or less screenshotting part of my buffer)

like it would be useful to have a function that goes:
void U8G2::getXBM(u8g2_uint_t x, u8g2_uint_t y, u8g2_uint_t w, u8g2_uint_t h, uint8_t *bitmap)

*answer provided by oliver in the reply section

**Comment by v1993** [comment, marked as answer]
https://github.com/olikraus/u8g2/wiki/u8g2reference#direct-access-buffer-api looks like what you want. I guess it's pretty obvious that you'll also want to use full-screen mode instead of paging one in this situation.

**Reply by stryxter** [reply]
this wont work on my lcd because my lcd doesnt support U8x8 memory structure

**Reply by olikraus** [reply]
Back reference: https://forum.arduino.cc/t/u8g2-screenshot/1104256/10

> this wont work on my lcd because my lcd doesnt support U8x8 memory structure

Well, as mentioned, the memory structure is the same as described in the ST7920 datasheet.
Assume you have x and y position provided, then the pixel value should be
``` C++
uint8_t *p = u8g2.getBufferPtr();
uint8_t v = p[x/8+y*u8g2.getBufferTileWidth()];
if ( v & (1<<(x&7))) {
  // pixel is set 
} else {
  // pixel is not set
}
```
I actually made a mistake in Arduino Forum, mentioning the wrong function.

> void U8G2::getXBM(u8g2_uint_t x, u8g2_uint_t y, u8g2_uint_t w, u8g2_uint_t h, uint8_t *bitmap)

Difficult, because the result might be very large (probably more the 4 times larger than the original data).
Usually this will not fit into any Arduino Uno. Writing to "serial" seems to be more suitable and will work with all displays and controllers.

**Reply by stryxter** [reply]
if you would like to, could you please give a detailed explanation on how the entire sketch works, like what does uint8_t *p and how  this helps...

**Reply by olikraus** [reply]
`uint8_t *p` defines a pointer to a memory location of size byte. It will be used later as an array of bytes.
